File: recipes/python/assets/get_vmware_assets.py
## The script can be run with Python 3.5 or higher version. 
## The script requires 'requests' library to make the API calls. The library can be installed using the command: pip install requests.

import argparse
import requests
import login.login_api as login_api

def get_usage():
    return ("\nThe command should be run from the parent directory of the 'assets' directory:\n"
            "python -Wignore -m assets.get_vmware_assets -nbserver <server> -username <username> -password <password> "
            "-domainName <domainName> -domainType <domainType> [-assetType <vm|vmGroup>] [-assetsFilter <filter criteria>]\n"
            "Optional arguments:\n"
            "assetType - vm or vmGroup. Default is 'vm'.\n"
            "assetsFilter - OData filter to filter the returned assets (VMs or VM Groups). If not specified, returns all the assets.\n")

parser = argparse.ArgumentParser(usage = get_usage())
parser.add_argument('-nbserver', required=True)
parser.add_argument('-username', required=True)
parser.add_argument('-password', required=True)
parser.add_argument('-domainName', required=True)
parser.add_argument('-domainType', required=True)
parser.add_argument('-assetType', default="vm", choices=['vm', 'vmGroup'])
parser.add_argument('-assetsFilter', default="")
args = parser.parse_args()

nbserver = args.nbserver
username = args.username
password = args.password
domainName = args.domainName
domainType = args.domainType
assetType = args.assetType
assetsFilter = args.assetsFilter

base_url = "https://" + nbserver + "/netbackup"
vm_assets_url = base_url + "/asset-service/workloads/vmware/assets"

default_sort = "commonAssetAttributes.displayName"

print("\nExecuting the script...")
jwt = login_api.perform_login(base_url, username, password, domainName, domainType)

print("\nGetting VMware {} assets...".format(assetType))

if assetType == "vm":
    assetTypeFilter = "(assetType eq 'vm')";
    print("Printing the following VM details: VM Display Name, Instance Id, vCenter, Protection Plan Names\n")
elif assetType == "vmGroup":
    assetTypeFilter = "(assetType eq 'vmGroup')";
    print("Printing the following VM group details: VM Group Name, VM Server, Filter Criteria, Protection Plan Names\n")

if assetsFilter != "":
    assetsFilter = assetsFilter + " and " + assetTypeFilter
else:
    assetsFilter = assetTypeFilter

headers = {'Authorization': jwt}

def get_vmware_assets():
    offset = 0
    next = True
    while next:
        queryparams = {'page[offset]':offset, 'filter':assetsFilter, 'sort':default_sort}
        assets_response = requests.get(vm_assets_url, headers=headers, params=queryparams, verify=False)
        assets = assets_response.json()

        if assets_response.status_code != 200:
            print("VMware Assets API returned status code: {}, response: {}\n".format(assets_response.status_code, assets_response.json()))
            raise SystemExit()

        print_assets(assets['data'])

        offset += assets['meta']['pagination']['limit']
        next = assets['meta']['pagination']['hasNext']

        if len(assets['data']) == 0:
            print("No assets returned.")

def print_assets(assets_data):
    for asset in assets_data:
        asset_attrs = asset['attributes']
        asset_common_attrs = asset_attrs['commonAssetAttributes']
        asset_protection_plans = []
        if "activeProtection" in asset_common_attrs :
            asset_protection_list = asset_common_attrs['activeProtection']['protectionDetailsList']
            for asset_protection in asset_protection_list:
                asset_protection_plans.append(asset_protection['protectionPlanName'])

        if assetType == "vm":
            print(asset_common_attrs['displayName'], asset_attrs['instanceUuid'], asset_attrs['vCenter'], asset_protection_plans, sep="\t")
        elif assetType == "vmGroup":
            print(asset_common_attrs['displayName'], asset_attrs['filterConstraint'], asset_attrs['oDataQueryFilter'], asset_protection_plans, sep="\t")


get_vmware_assets()

print("\nScript completed!\n")

File: recipes/go/assets/get_vmware_assets.go
//This script can be run using NetBackup 8.3 and higher.
//It gets the list of VMware assets in NetBackup (based on the given filter if specified, else returns all the VMware assets).

package main

import (
    "flag"
    "fmt"
    "log"
    "os"
    "strconv"
    "net/url"
    "net/http"
    "io/ioutil"
    "encoding/json"
    "utils"
)

var (
    nbserver            = flag.String("nbserver", "", "NetBackup Server")
    username            = flag.String("username", "", "User name for NetBackup API login")
    password            = flag.String("password", "", "Password for the given user")
    domainName          = flag.String("domainName", "", "Domain name")
    domainType          = flag.String("domainType", "", "Domain type")
    assetsFilter        = flag.String("assetsFilter", "", "Filter string (odata format) to filter the assets")
)

const usage = "\n\nUsage: go run ./get_vmware_assets.go -nbserver <NetBackup server> -username <username> -password <password> [-domainName <domainName>] [-domainType <domainType>] [-assetsFilter <filter>]\n\n"

func main() {
    // Print usage
    flag.Usage = func() {
        fmt.Fprintf(os.Stderr, usage)
        os.Exit(1)
    }

    // Read command line arguments
    flag.Parse()

    if len(*nbserver) == 0 {
        log.Fatalf("Please specify the name of the NetBackup Server using the -nbserver option.\n")
    }
    if len(*username) == 0 {
        log.Fatalf("Please specify the username using the -username option.\n")
    }
    if len(*password) == 0 {
        log.Fatalf("Please specify the password using the -password option.\n")
    }

    httpClient := apihelper.GetHTTPClient()
    jwt := apihelper.Login(*nbserver, httpClient, *username, *password, *domainName, *domainType)

    vmwareAssetsApiUrl := "https://" + *nbserver + "/netbackup/asset-service/workloads/vmware/assets"
    defaultSort := "commonAssetAttributes.displayName"
    assetTypeFilter := "(assetType eq 'vm')"

    req, err := http.NewRequest("GET", vmwareAssetsApiUrl, nil)

    if err != nil {
        fmt.Printf("Making new HTTP request failed with error: %s\n", err)
        panic("Script failed.")
    }

    req.Header.Add("Authorization", jwt)
    pageLimit := 100
    offset := 0
    next := true
    params := url.Values{}

    if assetsFilter != nil {
      filter := ""
      if *assetsFilter != "" {
        filter = *assetsFilter + " and " + assetTypeFilter
      } else {
          filter = assetTypeFilter
      }
      params.Add("filter", filter)
    }

    params.Add("sort", defaultSort)
    params.Add("page[offset]", strconv.Itoa(offset))
    params.Add("page[limit]", strconv.Itoa(pageLimit))

    fmt.Println("\nGetting VMware assets...")
    fmt.Println("Printing the following asset details: Display Name, VM InstanceId, vCenter, Protection Plan Names\n")

    for next {
      req.URL.RawQuery = params.Encode()
      resp, err := httpClient.Do(req)

      if err != nil {
          fmt.Printf("Get VMware Assets failed with error: %s\n", err)
          panic("Script failed.")
      } else {
          respJson, _ := ioutil.ReadAll(resp.Body)
          if resp.StatusCode == 200 {
            var respPayload interface{}
            json.Unmarshal(respJson, &respPayload)
            respData := respPayload.(map[string]interface{})
            assetsData := respData["data"].([]interface{})
            printAssetDetails(assetsData)
            next = respData["meta"].(map[string]interface{})["pagination"].
                (map[string]interface{})["hasNext"].(bool)
          } else {
            fmt.Println(string(respJson))
            next = false
          }
      }
      offset, _ = strconv.Atoi(params["page[offset]"][0])
      params["page[offset]"][0] = strconv.Itoa(offset + pageLimit)
    }

    fmt.Println("\nScript completed.\n")
}

func printAssetDetails(assets []interface{}) {
  for _, asset := range assets {
    assetAttrs := asset.(map[string]interface{})["attributes"].(map[string]interface{})
    assetCommonAttrs := assetAttrs["commonAssetAttributes"].(map[string]interface{})
    displayName := assetCommonAttrs["displayName"]
    instanceId := assetAttrs["instanceUuid"]
    vCenter := assetAttrs["vCenter"]

    var protectionPlans []string
    if activeProtections, protected := assetCommonAttrs["activeProtection"]; protected {
      protectionDetailsList := activeProtections.(map[string]interface{})["protectionDetailsList"].([]interface{})

      for _, protectionDetails := range protectionDetailsList {
        protectionPlans = append(protectionPlans, protectionDetails.
                (map[string]interface{})["protectionPlanName"].(string))
      }
    }
    fmt.Printf("%s\t%s\t%s\t%v\n", displayName, instanceId, vCenter, protectionPlans)
  }

}
